impl/sql/dberrors.go

```diff
@@ -0,0 +1,34 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import (
+	"errors"
+
+	"github.com/VividCortex/mysqlerr"
+	"github.com/go-sql-driver/mysql"
+)
+
+// IsDeadlock returns true if the error reports that a deadlock was found when trying to get lock.
+func IsDeadlock(err error) bool {
+	if err == nil {
+		return false
+	}
```

Comment:
```golang
func IsDeadlock(err error) bool {
	var mysqlErr *mysql.MySQLError
	return errors.As(err, &mysqlErr) && mysqlErr.Number == mysqlerr.ER_LOCK_DEADLOCK
}
```
?

Reply:
As panics if err is nil

Reply:
I checked, it doesn't. It panics if the second parameter is nil.

Reply:
https://play.golang.org/p/Mip5AL02_c8